07-map.slide

```diff
@@ -2,19 +2,148 @@ Maps in Go
 22 Feb 2017
 Tags: edmontongo, golang, workshop
 
+Lukasz Rozycki
+[[https://lukaszr.com]]
+
 * Maps
 
-composite literal for creating maps
+- Hash Table, Dictionary
+- Mapping from key to a value
+- Fast lookups, adds, and deletes
+- A Go map type looks like this:
+ map[KeyType]ValueType
+- Maps are not safe for concurrent use
+
+* What can be a key?
+
+- Any type that is comparable
+ ==    equal
+ !=    not equal
+- boolean, numeric, string, pointer
+- interface type, channel
+- struct or array containing above
+- Not slice, map, nor function
+
+* Creating a map 
+- composite literal (note trailing comma)
+.play 07-map/compositeliteral.go /START/,/END/
+- empty map
+.play 07-map/emptyliteral.go /START/,/END/
+
+* Lookup
+- index expression
+.play 07-map/lookup.go /START/,/END/
+- zero value
+.play 07-map/zerovalue.go /START/,/END/
+- what about Baby?
+.play 07-map/commaok1.go /START/,/END/
+
+* Lookup
+- "comma ok" idiom
+.play 07-map/commaok2.go /START/,/END/
+- concise notation
+.play 07-map/commaok3.go /START/,/END/
+
+* Lookup
+- ignore age
+.play 07-map/commaok4.go /START/,/END/
+- why underscore?
+.play 07-map/commaok5.go /START/,/END/
+
+* Addition
+- index expression
+.play 07-map/put.go /START/,/END/
+- read and modify
+.play 07-map/put2.go /START/,/END/
+
+* Deletion
+- delete built-in function
+ func delete(m map[Type]Type1, key Type)
+.play 07-map/delete.go /START/,/END/
+- delete no-op
+.play 07-map/delete2.go /START/,/END/
+
+* Wrong types
+- get wrong type
+.play 07-map/lookup3.go /START/,/END/
+- add wrong type
+.play 07-map/put3.go /START/,/END/
+- delete wrong type
+.play 07-map/delete3.go /START/,/END/
+
+* Set with map
+- create
+ set := map[int]bool{}
+- add to a aset
+ set[3] = true
+- remove from set
+ delete(set, 7)
+- check if set contains
+ if(set[13]) ...
+
+* Iteration
+- for range
+.play 07-map/for.go /START/,/END/
+- order is random
+.play 07-map/for2.go /START/,/END/
+
+* Iteration
+- blank value
+.play 07-map/for3.go /START/,/END/
+- key and value
+.play 07-map/for4.go /START/,/END/
+- blank key
+.play 07-map/for5.go /START/,/END/
+
+* Map is a reference type
+- zero value
+.play 07-map/nil.go /START/,/END/
+- add
+.play 07-map/nil2.go /START/,/END/
+- retrieve
+.play 07-map/nil3.go /START/,/END/
+- delete (no-op)
+.play 07-map/nil4.go /START/,/END/
+
+* Map is a reference type
+- passing to a function
+.code 07-map/ref.go /FUNC_S/,/FUNC_E/
+.play 07-map/ref.go /START/,/END/
+- assigning to a second variable
+.play 07-map/ref2.go /START/,/END/
+
+* Exercises
+
+- word count from A Tour of Go [[https://tour.golang.org/moretypes/23][Link]]
+ func CountWords(s string) map[string]int
+ use strings.Fields to split string
```

Comment:
strings.Fields returns a slice, which isn't covered until the next section. maybe a different example?